### PR TITLE
[Misc] Remove non-test uses of commons-lang3

### LINKED ISSUE #2100: Remove non-test uses of commons-lang3
Iceberg should avoid unnecessary dependencies, like commons-*. A few dependencies on commons have snuck in:

```
[blue@work iceberg]$ ack 'org.apache.commons'
flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java
29:import org.apache.commons.lang3.ArrayUtils;

versions.props
13:org.apache.commons:commons-lang3 = 3.9

pig/src/main/java/org/apache/iceberg/pig/IcebergPigInputFormat.java
30:import org.apache.commons.lang3.SerializationUtils;

spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java
28:import org.apache.commons.lang3.ArrayUtils;
```

We should remove those and move the dependency to testCompile.

### PR DESCRIPTION
**Pull request by pan3793**
close #2100

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/util/ArrayUtil.java [hunk 1]
```diff
@@ -59,4 +68,221 @@ private ArrayUtil() {
...
+      return null;
+    } else if (array.length == 0) {
+      return EMPTY_BOOLEAN_ARRAY;
+    }
+    final boolean[] result = new boolean[array.length];
+    for (int i = 0; i < array.length; i++) {
+      result[i] = array[i].booleanValue();
+    }
+    return result;
+  }
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/util/ArrayUtil.java:92]: If the code for these comes from `org.apache.commons:commons-lang3`, it should be credited as such some where in the file.

Comment (pan3793) [reply core/src/main/java/org/apache/iceberg/util/ArrayUtil.java:94]: Thanks for review. Updated the comments.

File: core/src/main/java/org/apache/iceberg/util/SerdeUtil.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+   *
+   * <p>The stream passed in is not buffered internally within this method.
+   * This is the responsibility of your application if desired.</p>
+   *
+   * @param obj          the object to serialize to bytes, may be null
+   * @param outputStream the stream to write to, must not be null
+   * @throws IllegalArgumentException if {@code outputStream} is {@code null}
+   * @throws UncheckedIOException     (runtime) if the serialization fails
+   */
+  public static void serialize(final Serializable obj, final OutputStream outputStream) {
```
Comment (rdsr) [inline core/src/main/java/org/apache/iceberg/util/SerdeUtil.java:71]: The `iceberg-mr` module also defines these methods as part of `org.apache.iceberg.mr.SerializationUtil` class. Though changing that may require some effort, maybe we can file a ticket to investigate and remove the duplication.

Comment (pan3793) [reply core/src/main/java/org/apache/iceberg/util/SerdeUtil.java:71]: Thanks for remind. I will merge `SerializationUtil` in to `SerdeUtil`.

File: core/src/main/java/org/apache/iceberg/util/ArrayUtil.java [hunk 3]
```diff
@@ -92,6 +94,8 @@ private ArrayUtil() {
   }
 
   /**
+   * <p>The code from `org.apache.commons:commons-lang3`
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/util/ArrayUtil.java:97]: Nit: I don't think you need an initial `<p>` tag.

What I've seen elsewhere in the code base is that the first line is written without `<p>`, and then subsequent paragraphs have a `<p>` tag between them where the `<p>` tag is on its own line. We tend to not use closing `</p>` tags as the java doc specification doesn't require them.

So here we would have
```java
/**
 * This code is borrowed from `org.apache.commons:commons-lang3`.
 * <p>
 * Converts an array of object Bytes to primitives.
 * <p>
 */
```

Comment (pan3793) [reply core/src/main/java/org/apache/iceberg/util/ArrayUtil.java]: make sense. updated

**Comment by pan3793** [comment]
@rdblue Could you please review this PR ?

**Comment by rdsr** [comment]
LG!. I realize that you might have to update the LICENSE file which lists all the code we have borrowed from different project.

**Comment by pan3793** [comment]
> LG!. I realize that you might have to update the LICENSE file which lists all the code we have borrowed from different project.

Updated

**Review by rdsr — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/util/ArrayUtil.java [hunk 4]
```diff
@@ -59,4 +68,239 @@ private ArrayUtil() {
...
+    }
+    final double[] result = new double[array.length];
+    for (int i = 0; i < array.length; i++) {
+      result[i] = array[i].doubleValue();
+    }
+    return result;
+  }
+
+  /**
+   * This code is borrowed from `org.apache.commons:commons-lang3`.
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/ArrayUtil.java:241]: Could you move this comment so it is the last paragraph? It is important to note, but not as important as the usage docs.

Comment (pan3793) [reply core/src/main/java/org/apache/iceberg/util/ArrayUtil.java]: Updated.

File: LICENSE [hunk 5]
```diff
@@ -299,3 +299,13 @@ This product includes code from Delta Lake.
...
+
+--------------------------------------------------------------------------------
+
+This product includes code from Apache Commons.
+
+* Core ArrayUtil.
+
+Copyright: 2020 The Apache Software Foundation
+Home page: https://commons.apache.org/
+License: https://www.apache.org/licenses/LICENSE-2.0
```
Comment (rdblue) [inline LICENSE:311]: This needs to be added to all of the runtime LICENSE files as well, like spark3-runtime/LICENSE. Can you add it to all of the runtime license files, please?

File: core/src/main/java/org/apache/iceberg/util/SerdeUtil.java [hunk 6]
```diff
@@ -28,14 +28,14 @@
 import java.nio.charset.StandardCharsets;
 import java.util.Base64;
 
-public class SerializationUtil {
+public class SerdeUtil {
 
-  private SerializationUtil() {
+  private SerdeUtil() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/SerdeUtil.java:33]: Are these change related? Why the rename?

File: LICENSE (hunk 5 above)
Comment (pan3793) [reply LICENSE:311]: Sure.

File: core/src/main/java/org/apache/iceberg/util/SerdeUtil.java (hunk 6 above)
Comment (pan3793) [reply core/src/main/java/org/apache/iceberg/util/SerdeUtil.java:33]: Not realized there is a `SerializationUtil` in `mr` module at begin, so create a new `SerdeUtil` in `core`. I will revert the name changed.

**Comment by pan3793** [comment]
CI failed because of #2098

File: core/src/main/java/org/apache/iceberg/util/SerializationUtil.java [hunk 7]
```diff
@@ -35,7 +35,7 @@ private SerializationUtil() {
 
   public static byte[] serializeToBytes(Object obj) {
     try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+         ObjectOutputStream oos = new ObjectOutputStream(baos)) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/SerializationUtil.java:38]: Can you remove the whitespace only changes in this file?

**Comment by rdblue** [comment]
The only remaining issues were nits, so I merged this. Thanks @pan3793!
